### PR TITLE
add server timeout callbacks

### PR DESCRIPTION
**Pull request by fpliger**

### CODE REVIEW

File: bokeh/document.py [hunk 1]
```diff
@@ -523,6 +542,37 @@ def remove_periodic_callback(self, callback):
...
+    def remove_timeout_callback(self, callback):
+        ''' Remove a callback added earlier with add_timeout_callback()
+
+            Throws an error if the callback wasn't added
+
+        '''
+        self.remove_session_callback(callback)
+
+
+    def remove_session_callback(self, callback):
```
Comment (fpliger) [inline bokeh/document.py]: @bokeh/core I'm doing this repetition (having different remove for timeout and periodic and a remove session) for the sake expressiveness.. but feel if could be better. Any suggestions? I'm thinking to at least turn `remove_session_callback` private..

File: bokeh/client/session.py [hunk 2]
```diff
@@ -306,11 +323,23 @@ def _document_changed(self, event):
...
-        if isinstance(event, SessionCallbackRemoved):
-            self._remove_periodic_callback(event.callback)
+        elif isinstance(event, SessionCallbackRemoved):
+            if isinstance(event.callback, PeriodicCallback):
+                self._remove_periodic_callback(event.callback)
+            elif isinstance(event.callback, TimeoutCallback):
+                self._remove_timeout_callback(event.callback)
+            else:
+                raise ValueError("Expected callback of type PeriodicCallback or TimeoutCallback, got: %s" % event.callback)
+
```
Comment (fpliger) [inline bokeh/client/session.py]: @bokeh/core  would like to move all these actions outside... I'm debated between moving those to a `session.act_on_event` method or move each action logic to it's own event implementation (in this case it'd take the session). 

My preference is for the later option.. opinions?

Comment (havocp) [inline bokeh/client/session.py]: It bothers me to have knowledge of Session inside `document.py` or inside the events. This is a multiple-dispatch problem I guess (http://python-3-patterns-idioms-test.readthedocs.org/en/latest/MultipleDispatching.html ). One solution could be this (untested but hopefully you see what I mean):

```
class DocumentChangedEvent(object):
    def dispatch(self, receiver):
        if hasattr(receiver, '_document_changed'):
          receiver._document_changed(self)

class SessionCallbackAdded(DocumentChangedEvent):
    def dispatch(self, receiver):
        super(SessionCallbackAdded, self).dispatch(receiver)
        if hasattr(receiver, '_session_callback_added'):
          receiver._session_callback_added(self)
```

Receiver would have to attach like this I guess:
```
 document.on_change(lambda event: event.dispatch(self))
```

That could maybe be a convenience method on Document, like:
```
class Document:
  def on_change_dispatch_to(self, receiver):
     self.on_change(lambda event: event.dispatch(receiver))
```

?

File: bokeh/document.py (hunk 1 above)
Comment (havocp) [inline bokeh/document.py]: Agree I think I'd make `remove_session_callback` private too.

File: bokeh/client/session.py (hunk 2 above)
Comment (bryevdv) [inline bokeh/client/session.py]: Agreed this seems like MD

Comment (havocp) [inline bokeh/client/session.py]: if you go the event.dispatch route you might want to make it its own PR since you'd have to do all the existing events as well as your new ones.

Comment (fpliger) [inline bokeh/client/session.py]: yup, +1 for multiple-dispatch and agree it's a bit over the scope of this PR. Living it for a following one.

In this case I'll leave as is for now if there are no objections

**Comment by havocp** [comment]
Merged, lgtm
